content/blogposts/2024/gp-exploring-data-visualization-with-d3-cody.md

```diff
@@ -0,0 +1,326 @@
+* **Framework Support**: As a JavaScript library, D3.js works with all JavaScript frameworks, making it widely used by a highly active community of users. 
+## Getting started with D3.js
+
+Now that you have learned about data visualization and the features of D3.js, it is time to see how it works in our application. In this article, we will be working on a Vanilla HTML environment. You can load D3.js from a CDN like so:
+
+```html
+<script type="module">
+  import * as d3 from "https://cdn.jsdelivr.net/npm/d3@7/+esm";
+</script>
+```
+
+## Working principle
+
+D3.js allows developers to bind arbitrary data to the DOM, which makes it easy to manipulate and update data in real-time. In this section, we will go through the basic logic of selecting elements to be visualized, modifying the elements with custom styling, and joining the data. 
+
+### Selecting elements
+
+The concept of selecting elements in D3.js isn’t so strange, especially from a JavaScript background. Selecting elements basically entails targeting the particular element you want to modify and visualize. 
+
+Suppose you want to visualize elements with the class `chartbox`, you simply use the `.chartbox` class selector, or a `div` element selector to select a `div` element. D3.js provides two methods of selecting elements; the `d3.select()` method and the `d3.selectAll` method. The `d3.select` method selects only the first matching element while `d3.selectAll` selects all matching instances in the DOM(from top to bottom). Consider the practical example below to learn more about how the select methods work:
+
+```html
+<div>
+  <h1>Cody by Sourcegraph is the best AI coding assistant</h1>
+  <h1>Cody by Sourcegraph is the best AI coding assistant</h1>
+  <h1>Cody by Sourcegraph is the best AI coding assistant</h1>
+</div>
+```
+
+Then I select the `h1` element with the `.select()` method:
+
+```javascript
+d3.select("h1").style("color", "red");
+```
+
+Here’s the result:
+
+<Figure
+  src="https://storage.googleapis.com/sourcegraph-assets/blog/gp-exploring-data-visualization-with-d3-cody/image_001.png"
+/>
+
+We see that only the first `h1` element is affected because we used the `.select()` method. We can toggle the codebase to target all matching instances of `h1` using the `.selectAll()`. Rather than manually make the changes, we can allow Cody to make further edits or manipulation by simply **telling it to do so**, since we already established a context in our codebase.
+
+<Figure
+  src="https://storage.googleapis.com/sourcegraph-assets/blog/gp-exploring-data-visualization-with-d3-cody/image_002.gif"
+  caption="Using Cody's Edit Command"
+/>
+
+And there! Come on, how interesting is this? I tell Cody to _change the method to .selectAll and the color to blue_, and then click on **accept**(since it’s precisely the change I want to make). Here’s the result:
```

Comment:
```suggestion
And there! Come on, how interesting is this? I tell Cody to _change the method to .selectAll and the color to blue_, and then click on **accept** (since it’s precisely the change I want to make). Here’s the result:
```